File: src/scrolls.js
// Consumable scrolls the player can run to provide flat stats — letting the gem/enchant/item budget
// shift toward threat while the scroll covers some avoidance/defense. TBC rank-V values (the scrolls
// stack with Kings/MotW and each other). Primary-stat scrolls feed `buffs` (so Kings' +10% applies,
// exactly as in-game); Scroll of Protection's armor feeds `flatArmor` so Toughness — which only
// boosts armor FROM ITEMS — does NOT inflate it.
export const SCROLLS = {
  agility: { name: 'Scroll of Agility V', stat: 'agility', value: 20 },     // -> dodge
  strength: { name: 'Scroll of Strength V', stat: 'strength', value: 20 },  // -> block value
  intellect: { name: 'Scroll of Intellect V', stat: 'intellect', value: 20 }, // -> spell crit/mana
  protection: { name: 'Scroll of Protection V', stat: 'armor', value: 301, flat: true }, // +301 armor
};

// Sum a list of scroll keys into { buffs: {primary stats}, flatArmor }. Unknown keys are ignored.
export function scrollStats(keys = []) {
  const buffs = {};
  let flatArmor = 0;
  for (const key of keys) {
    const s = SCROLLS[key];
    if (!s) continue;
    if (s.flat) flatArmor += s.value;
    else buffs[s.stat] = (buffs[s.stat] || 0) + s.value;
  }
  return { buffs, flatArmor };
}

File: addon/TankadinGearSim/engine/Scrolls.lua
-- Scrolls — Lua port of src/scrolls.js:scrollStats. Sums a list of consumable-scroll keys into a
-- { buffs = {primary stats}, flatArmor } block: primary-stat scrolls ride the buff block (so Kings'
-- +10% applies), Scroll of Protection's armor is flat (bypasses Toughness). Data lives in
-- engine/ScrollsData.lua. Parity-checked by test/lua/solver_parity.lua.

local ADDON, ns = ...
ns.engine = ns.engine or {}
local D = ns.engine.ScrollsData

local Scrolls = {}
ns.engine.Scrolls = Scrolls

-- Sum scroll keys -> { buffs, flatArmor }. Unknown keys are ignored. Mirrors scrolls.js:scrollStats.
function Scrolls.scrollStats(keys)
  keys = keys or {}
  local buffs = {}
  local flatArmor = 0
  for _, key in ipairs(keys) do
    local s = D.SCROLLS[key]
    if s then
      if s.flat then flatArmor = flatArmor + s.value
      else buffs[s.stat] = (buffs[s.stat] or 0) + s.value end
    end
  end
  return { buffs = buffs, flatArmor = flatArmor }
end

return Scrolls
